core/src/main/java/org/infinispan/statetransfer/StateTransferManagerImpl.java

```diff
@@ -179,37 +183,56 @@ private CacheTopology addPartitioner(CacheTopology cacheTopology) {
             unionCH, cacheTopology.getPhase(), cacheTopology.getActualMembers(), cacheTopology.getMembersPersistentUUIDs());
    }
 
-   private void doTopologyUpdate(CacheTopology newCacheTopology, boolean isRebalance) {
+   private CompletionStage<Void> doTopologyUpdate(CacheTopology newCacheTopology, boolean isRebalance) {
       CacheTopology oldCacheTopology = distributionManager.getCacheTopology();
 
       int newTopologyId = newCacheTopology.getTopologyId();
       if (oldCacheTopology != null && oldCacheTopology.getTopologyId() > newTopologyId) {
-         throw new IllegalStateException("Old topology is higher: old=" + oldCacheTopology + ", new=" + newCacheTopology);
+         throw new IllegalStateException(
+               "Old topology is higher: old=" + oldCacheTopology + ", new=" + newCacheTopology);
       }
 
       if (trace) {
          log.tracef("Installing new cache topology %s on cache %s", newCacheTopology, cacheName);
       }
 
-      // No need for extra synchronization here, since LocalTopologyManager already serializes topology updates.\
+      // No need for extra synchronization here, since LocalTopologyManager already serializes topology updates.
       if (newCacheTopology.getMembers().contains(rpcManager.getAddress())) {
-         if (!distributionManager.getCacheTopology().isConnected() || !distributionManager.getCacheTopology().getMembersSet().contains(rpcManager.getAddress())) {
-            if (trace) log.tracef("This is the first topology %d in which the local node is a member", newTopologyId);
+         if (!distributionManager.getCacheTopology().isConnected() ||
+             !distributionManager.getCacheTopology().getMembersSet().contains(rpcManager.getAddress())) {
+            if (trace)
+               log.tracef("This is the first topology %d in which the local node is a member", newTopologyId);
             inboundInvocationHandler.setFirstTopologyAsMember(newTopologyId);
          }
       }
 
       // handle the partitioner
-      newCacheTopology = addPartitioner(newCacheTopology);
+      CacheTopology partitionerCacheTopology = addPartitioner(newCacheTopology);
       int newRebalanceId = newCacheTopology.getRebalanceId();
       CacheTopology.Phase phase = newCacheTopology.getPhase();
 
-      // TODO: these should be non blocking at some point
-      CompletionStages.join(cacheNotifier.notifyTopologyChanged(oldCacheTopology, newCacheTopology, newTopologyId, true));
+      return cacheNotifier.notifyTopologyChanged(oldCacheTopology, partitionerCacheTopology, newTopologyId, true)
+                          .thenCompose(ignored -> {
```

Comment:
Nitpick: Extra braces

Reply:
I put those intentionally, without the braces the parameters would go too much to the right IMO